### ISSUE TITLE
Kestrel uses exception for control flow when handling the `"Connection": "close"` header.

### ISSUE THREAD

**Issue by Tomius** [dotnet/aspnetcore#41206]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

When Kestrel receives an http request with the `"Connection": "close"` header, then during the process of handling it a `SocketException` or `ObjectDisposedException` will be thrown and then caught. While the request will be handled correctly (at least I believe so), the exception in the hot path is undesirable, because:
* Throwing and catching the exception is definitely slower than inspecting the error code.
* If the application is recording a list of all the exceptions happening in a websever, this creates a significant noise (in my company there are 100 times as many exceptions coming from this source than all the other exception sources combined). And the recorded exception is hard understand or to link with the real cause.
* The thrown exception and stack trace is not always the same:
On Windows it's usually:
```
System.Net.Sockets.SocketException (995): The I/O operation has been aborted because of either a thread exit or an application request.
   at Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.Internal.SocketAwaitableEventArgs.<GetResult>g__ThrowSocketException|5_0(SocketError e)
   at Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.Internal.SocketAwaitableEventArgs.GetResult(Int16 token)
   at Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.Internal.SocketConnection.DoReceive()
```
In linux it's usually:
```
System.Net.Sockets.SocketException (125): Operation canceled at
Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.Internal.SocketAwaitableEventArgs.<GetResult>g__ThrowSocketException|7_0(SocketError e) at
Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.Internal.SocketAwaitableEventArgs.GetResult() at
Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.Internal.SocketConnection.ProcessReceives() at
Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.Internal.SocketConnection.DoReceive()
```
But sometimes the following exception is thrown instead. It happens about 5% of the times in production in my setup. Interestingly if I put a breakpoint into `SocketAsyncEventArgs.SocketError.get`, this exception will be thrown instead of the SocketException 100% of the times.
```
System.ObjectDisposedException: Cannot access a disposed object.
Object name: 'System.Net.Sockets.Socket'.
   at System.Net.Sockets.Socket.ThrowObjectDisposedException()
   at System.Net.Sockets.Socket.ReceiveAsync(SocketAsyncEventArgs e, CancellationToken cancellationToken)
   at System.Net.Sockets.Socket.ReceiveAsync(SocketAsyncEventArgs e)
   at Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.Internal.SocketReceiver.ReceiveAsync(Socket socket, Memory`1 buffer)
   at Microsoft.AspNetCore.Server.Kestrel.Transport.Sockets.Internal.SocketConnection.DoReceive()
```
* It's a frequent use-case, for consul http health checks, that always use the `"Connection": "close"` header. This means that in my company, 100% of our services in production are affected by this, with an average of 100 exceptions / application / minute.

### Steps To Reproduce

- Create a hello world Kestrel app with `dotnet new webapp`
    - (Make sure it uses Kestrel and not HttpSys, add `.UseKestrel()` if necesssary)
- Add a breakpoint into `SocketAwaitableEventArgs.GetResult.ThrowSocketException`
- Write a simple client:
```
using HttpClient http = new HttpClient();
http.DefaultRequestHeaders.Connection.Add("close");
await http.GetAsync(url);
```

### .NET Version

I could reproduce it with asp net core 2.1, 3.1 and 6..0

### Anything else?

Where is the exception coming from?
==========================

I know that the exception is thrown here:
https://github.com/dotnet/aspnetcore/blob/c85baf8db0c72ae8e68643029d514b2e737c9fae/src/Servers/Kestrel/Transport.Sockets/src/Internal/SocketAwaitableEventArgs.cs#L49

And it is caught here:
https://github.com/dotnet/aspnetcore/blob/4931b1929188349b438575803bcec889a9a7d190/src/Servers/Kestrel/Transport.Sockets/src/Internal/SocketConnection.cs#L198

But I wasn't able to figure out where the `SocketError` inside the `SocketAwaitableEventArgs` is set.

**Comment by Tratcher** [comment]
Related? https://github.com/dotnet/aspnetcore/pull/38094 @sebastienros

**Comment by sebastienros** [comment]
@Tratcher exactly that, merge the PR ;)

**Comment by adityamandaleeka** [comment]
Fixed in #38094
